src/language/pattern-types.md

```diff
@@ -0,0 +1,333 @@
+{% endcomment %}
+
+Logical-or patterns are useful for having multiple cases share a body in switch
+expressions or statements.
+
+```dart
+var isPrimary = switch (color) {
+  Color.red || Color.yellow || Color.blue => true,
+  _ => false
+};
+```
+
+Switch statements can have multiple cases share a body without using logical-or
+patterns, but they are still uniquely useful for allowing multiple cases to share
+a guard: 
+
+```dart
+switch (shape) {
+  case Square(size: var s) || Circle(size: var s) when s > 0:
+    print('Non-empty symmetric shape');
+}
+```
+
+## Map
+
+`{"key": subpattern1, someConst: subpattern2}`
+
+Map patterns match values that implement [`Map`][], and then recursively 
+match its subpatterns against the map’s keys to destructure them.
+
+Map patterns don't require the pattern to match the entire map. If a map has
+extra keys that te pattern doesn't destructure, it can still match.
+
+## Null-assert	
+
+`subpattern!`
+
+Null-assert patterns permit non-null values to flow through, but [throw][] if the
+matched value is null. 
+
+```dart
+List<String?> row = // ...
+
+// If the first column is 'user', we expect to have a name after it.
+switch (row) {
+  case ['user', var name!]:
+    // name is a non-nullable string here.
+}
+```
+
+## Null-check	
+
+`subpattern?`
+
+Null-check patterns match if the value is not null, and then match the inner
+pattern against that same value.
+
+
+```dart
+String? maybeString = // ...
+switch (maybeString) {
+  case var s?:
+    // s has type non-nullable String here.
+}
+```
+
+## Object
+
+`SomeClass(x: subpattern1, y: subpattern2)`
+
+Object patterns checks the matched value against a given named type to destructure
+data using the getters the object's class already defines. They are [refuted][]
+if the value doesn't have the same type.
+
+The getter name can be omitted and inferred from the [variable pattern](#variable)
+or [identifier pattern](#identifier) in the field subpattern:
+
+```dart
+var Point(:var x, :var y) = Point(1, 2);
+```
+
+Object patterns don't require the pattern to match the entire object.
+If an object has extra fields that the pattern doesn't destructure, it can still match.
+
+## Parenthesized
+
+`(subpattern)`
+
+Like parenthesized expressions, parentheses in a pattern let you control pattern
+precedence and insert a lower precedence pattern where a higher precedence one is
+expected.
+
+## Record
+
+`(subpattern1, subpattern2)`
+
+`(x: subpattern1, y: subpattern2)`
+
+Record patterns match a [record][] object and destructure its fields.
+If the value isn't a record with the same [shape][] as the pattern, the match
+fails. Otherwise, the field subpatterns are matched against the corresponding
+fields in the record.
+
+```dart
+// Variable:
+var (untyped: untyped, typed: int typed) = // ...
+var (:untyped, :int typed) = // ...
+
+switch (obj) {
+  case (untyped: var untyped, typed: int typed): // ...
+  case (:var untyped, :int typed): // ...
+}
```

Comment:
I would move the typed / untyped stuff to the variable patterns section, since it's not specific to how record patterns work.